man/man1/xtfs_chstatus.1
.TH xtfs_chstatus 1 "July 2011" "The XtreemFS Distributed File System" "XtreemFS Admin Tools"
.SH NAME
xtfs_chstatus \- changes the status flag of an OSD.
.SH SYNOPSIS
\fBxtfs_chstatus [ \fIoptions\fB ] \fI<osd_uuid> online|removed\fB
.br

.SH DESCRIPTION
.I xtfs_chstatus
displays and changes the status flag of an OSD. The status can be 0 (online), 1 (locked) and 2 (permanently dead/removed). The status flag is used by the scrubber to identify dead OSDs. Files and replicas on dead OSDs are removed by the scrubber tool.

.SH EXAMPLE USAGE
.TP
.B "xtfs_chstatus -dir pbrpc://localhost:32638 2e3d-424b-b3d2 removed
Changes the status of the OSD with UUID 2e3d-424b-b3d2 registered at the Directory Service 'pbrpc://localhost:32638' to 2 (removed)

.SH OPTIONS
.TP
\fB-c \fI<creds_file>
Path to a PKCS#12 credentials file (private key + certificate) to use for SSL authentication. Must be present when MRC URL starts with pbrpcs:// or pbrpcg://.
.TP
\fB-cpass \fI<creds_passphrase>
An optional passphrase to access the credentials file.
.TP
\fB-dirs \fI<uri(,uri...)>
Specifies the directory services to use (e.g. 'pbrpc://localhost:32638,pbrpc://myDir.com:32638') (comma-separated). If no URI is specified, URIs and security settings are taken from '/etc/xos/xtreemfs/default_dir'. In case of a secured URI ('pbrpcs://...' or 'pbrpcg://...'), it is necessary to also specify SSL settings.
.TP
\fB-h/--help
Show usage info.
.TP
\fB-t \fI<trusted_CAs>
Path to a PKCS#12 file containing a set of certificates from trusted certification authorities. These certificates will be used to authenticate the MRC. Must be present when MRC URL starts with pbrpcs:// or pbrpcg://.
.TP
\fB-tpass \fI<trusted_passphrase>
An optional passphrase to access the truststore file.
.TP
\fB---ssl-protocol \fI<ssl-protocol> 
SSL/TLS version to use: ssltls, tlsv1, tlsv11, tlsv12. 'ssltls' (default) accepts all versions, the others accept only the exact version they name. 'tlsv12' is available in JDK 7+ only. 'tlsv11' comes with JDK 6 or 7, depending on the vendor.

.SH "SEE ALSO"
.BR xtfs_cleanup (1),
.BR xtfs_mrcdbtool (1),
.BR xtfs_remove_osd (1),
.BR xtfs_scrub (1),
.BR

.SH AVAILABILITY
The xtfs_chstatus command is part of the XtreemFS-tools package and is available from \fIhttp://www.xtreemfs.org\fP.


CHANGELOG
26-OCT-2016: release 1.6.0 (Phantastic Pancakes)

  Changes:
    * Extended volume quota implementation to OSDs. This requires dumping and reimporting the MRC database during upgrade (see 'MRC Database Backups and Conversion' in the user guide).
      * Implemented hard quotas for volumes and user and groups per volume.
      * Added quota information to import and export function.
      * These changes supersede the previous implementation.
    * Fix for mount.xtreemfs to fully support mount options via /etc/fstab.
    * Introduced Java wrappers for the C++ version of libxtreemfs and made it the default Java implementation.
    * Improved Hadoop adapter to work with a broader range of systems and versions.
    * Added 'listen.port.bind_retries' server option that implements exponential backoff during service startup if the port is not free yet.
    * Added RabbitMQ OSD tracing policy (ID 6003).
    * Added Preferred UUID OSD selection policy (ID 1003).
    * Extended X509 authentication to allow user and group mapping.
    * Slimmed down and updated Mac OS X installer with stricter dependency checking and SIP awareness.
    * Converted the Java parts to Maven and made all components available as artifacts hosted on GitHub.
    * Cleanup of code base to exclude generated sources and previously shipped binaries.
    * Add non-interactive console support to command line tools.
    * Remove explicit support for SSLv3.
    * Many small stability improvements and fixes.

12-MAR-2015: release 1.5.1

  Changes:
    * Hadoop Adapter supports Hadoop-2.x and other applications running on the YARN platform (tested with Apache Spark and Apache Flink), new OSD selection policy that imitates the HDFS placement behavior.
    * Consistent adding and removing replicas for R/W replication.
    * xtfs_scrup now replaces failed replicas using the R/W replication policy.
    * Improvements of the SSL mode: the used SSL/TLS version is selectable, strict certificate chain checks are possible, the SSL code on client and server side was improved.
    * Support for PKCS12 certificate containers is remove from the Windows client (did not work in previous release).
    * mount.xtreemfs accepts all parameters to be passed in the form -o option=value. If XtreemFS is mounted via /etc/fstab, mount options are passed to the XtreemFS client.
    * Dockerfiles for the XtreemFS services are available at https://github.com/xtreemfs/xtreemfs-docker.
    * A Vagrantfile is available in the XtreemFS repository to ease development and testing for new contributors.
    * Fixed snapshots in the case of deleted files.
    * Snapshots can be deleted even after inactivating snapshots for a volume (issue 290).
    * Initial version of an LD_PRELOAD based client that bypasses FUSE (experimental).
    * Initial implementation of volume quotas (checked only while opening files by the MRC).
    * OSDs can report their health to the DIR (e.g. based on SMART values), which aggregates this information in the web interface. A new OSD selection policy sorts out unhealthy OSDs.
    * Improved automatic testing: introduced Travis-CI, migrated all JUnit tests to version 4, improved reliability of nightly integration tests.
    * Added new benchmark tool xtfs_benchmark that is based on the JAVA libxtreemfs.
    * Many small improvements and fixes.

12-MAR-2014: release 1.5 (Wonderful Waffles)

  Changes:
    * Improved Hadoop Adapter e.g., added support for multiple volumes and read and write buffer to speed up performance.
    * Support for SSDs: Support for multiple OSD storage threads to increase parallelism.
    * Status webpage as part of the DIR webinterface for replicated files to show current primary and backup replicas.
    * Multi-homing support. XtreemFS can now be made available for multiple networks and clients will pick the correct address automatically.
    * Support for multiple OSDs per machine (e.g. one for each disk) through the xtreemfs-osd-farm init.d script.
    * Fixed major issues in Read-Only and Read/Write file replication.
    * Reduced metadata cache default timeout from 120 to 10 seconds. Disable it completely with --metadata-cache-size=0.
    * Upgraded to Google Protobuffers 2.5 which improves performance.
    * Windows Client: Fixed several issues (e.g., 286, 299) and updated the CbFS driver to the latest version.
    * XtreemFS also compiles under buildroot now.
    * Fixed several issues around SSL (e.g., 263, 281).
    * Fixed many small issues e.g., reported volume capacity was inaccurate (issue 296).
      Other fixed issues: 267, 268, 272, 274, 276, 277, 278, 283, 287, 288, 292, 293, 294.
    * Improved general stability (e.g., see issues 297, 301).
    * xtfs_scrub now repairs replicas with an invalid checksum automatically.
    * Disabled deferred deletion of file data on the OSDs.
    * Improved test coverage of nightly tests and added missing unit tests and a Valgrind leak check for C++ unit tests.

12-NOV-2012: release 1.4 (Salty Sticks)

  Changes:
    * Improved stability: Fixed client crashes when timeouts occurred; fixed server crashes and issues with the R/W replication.
    * Full support for asynchronous writes: If a client is started with "--enable-async-writes", write() requests will be immediately acknowledged by the client and executed in the background. Outstanding writes are always flushed at close() or fsync*() events. This improves the write throughput, especially in case of connections with high latency.
    * Added Windows Client Beta which uses Eldos' Callback File System product instead of Dokan.
    * Re-wrote HDFS interface implementation: Use XtreemFS as replacement for HDFS in your Hadoop setup.
    * libxtreemfs for Java: Access XtreemFS directly from your Java application. See source files org.xtreemfs.common.libxtreemfs.{Client,Volume,FileHandle}.java for the interface.
    * Re-added Vivaldi support and added/improved its functionality:

java/servers/src/org/xtreemfs/utils/xtfs_chstatus.java
/*
 * Copyright (c) 2009-2011 by Bjoern Kolbeck, Zuse Institute Berlin
 *
 * Licensed under the BSD License, see LICENSE file for details.
 *
 */

package org.xtreemfs.utils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.xtreemfs.common.libxtreemfs.AdminClient;
import org.xtreemfs.common.libxtreemfs.ClientFactory;
import org.xtreemfs.common.libxtreemfs.Options;
import org.xtreemfs.foundation.SSLOptions;
import org.xtreemfs.foundation.logging.Logging;
import org.xtreemfs.foundation.pbrpc.Schemes;
import org.xtreemfs.foundation.pbrpc.client.RPCAuthentication;
import org.xtreemfs.foundation.util.CLIParser;
import org.xtreemfs.foundation.util.CLIParser.CliOption;
import org.xtreemfs.pbrpc.generatedinterfaces.DIR.ServiceStatus;
import org.xtreemfs.pbrpc.generatedinterfaces.GlobalTypes.PORTS;

/**
 * 
 * @author bjko
 */
public class xtfs_chstatus {

    public static void main(String[] args) {

        Logging.start(Logging.LEVEL_WARN);

        Map<String, CliOption> options = utils.getDefaultAdminToolOptions(false);
        List<String> arguments = new ArrayList<String>(1);
        CliOption oDir = new CliOption(
                CliOption.OPTIONTYPE.STRING,
                "directory service to use (e.g. 'pbrpc://localhost:32638'). If no URI is specified, URI and security settings are taken from '"
                        + DefaultDirConfig.DEFAULT_DIR_CONFIG + "'", "<uri>");
        oDir.urlDefaultPort = PORTS.DIR_PBRPC_PORT_DEFAULT.getNumber();
        oDir.urlDefaultProtocol = Schemes.SCHEME_PBRPC;
        options.put("dir", oDir);

        CLIParser.parseCLI(args, options, arguments);

        if (options.get(utils.OPTION_HELP).switchValue || options.get(utils.OPTION_HELP_LONG).switchValue) {
            usage(options);
            return;
        }

        if (arguments.size() > 2 || arguments.size() < 1)
            error("invalid number of arguments", options);

        String[] dirURLs = (options.get("dir").stringValue != null) ? options.get("dir").stringValue
                .split(",") : null;

        SSLOptions sslOptions = null;
        String[] dirAddrs = null;

        // parse security info if protocol is 'https'
        if (dirURLs != null) {
            int i = 0;
            boolean gridSSL = false;
            dirAddrs = new String[dirURLs.length];
            for (String dirURL : dirURLs) {

                // parse security info if protocol is 'https'
                if (dirURL.contains(Schemes.SCHEME_PBRPCS + "://")
                        || dirURL.contains(Schemes.SCHEME_PBRPCG + "://") && sslOptions == null) {
                    String serviceCredsFile = options.get(utils.OPTION_USER_CREDS_FILE).stringValue;
                    String serviceCredsPass = options.get(utils.OPTION_USER_CREDS_PASS).stringValue;
                    if(serviceCredsPass != null && serviceCredsPass.equals("-")) {
                    	serviceCredsPass = new String(System.console().readPassword("Enter credentials password: "));
                    }
                    String trustedCAsFile = options.get(utils.OPTION_TRUSTSTORE_FILE).stringValue;
                    String trustedCAsPass = options.get(utils.OPTION_TRUSTSTORE_PASS).stringValue;
                    if(trustedCAsPass != null && trustedCAsPass.equals("-")) {
                    	trustedCAsPass = new String(System.console().readPassword("Enter credentials password: "));
                    }
                    String sslProtocolString = options.get(utils.OPTION_SSL_PROTOCOL).stringValue;
                    if (dirURL.contains(Schemes.SCHEME_PBRPCG + "://")) {
                        gridSSL = true;
                    }

                    if (serviceCredsFile == null) {
                        System.out.println("SSL requires '-" + utils.OPTION_USER_CREDS_FILE
                                + "' parameter to be specified");
                        usage(options);
                        System.exit(1);
                    } else if (trustedCAsFile == null) {
                        System.out.println("SSL requires '-" + utils.OPTION_TRUSTSTORE_FILE
                                + "' parameter to be specified");
                        usage(options);
                        System.exit(1);
                    }

                    // TODO: support custom SSL trust managers
                    try {
                        sslOptions = new SSLOptions(serviceCredsFile, serviceCredsPass, SSLOptions.PKCS12_CONTAINER,
                                trustedCAsFile, trustedCAsPass, SSLOptions.JKS_CONTAINER, false, gridSSL,
                                sslProtocolString, null);
                    } catch (Exception e) {
                        System.err.println("unable to get SSL options, because:" + e.getMessage());
                        System.exit(1);
                    }
                }

                // add URL to dirAddrs
                if (dirURL.contains("://")) {
                    // remove Protocol information
                    String[] tmp = dirURL.split("://");
                    // remove possible slash
                    dirAddrs[i++] = tmp[1].replace("/", "");
                } else {
                    // remove possible slash
                    dirAddrs[i++] = dirURL.replace("/", "");
                }
            }

        }

        // read default settings
        if (dirURLs == null) {

            try {
                DefaultDirConfig cfg = new DefaultDirConfig();
                sslOptions = cfg.getSSLOptions();
                dirAddrs = cfg.getDirectoryServices();
            } catch (Exception e) {
                System.err.println("unable to get SSL options, because: " + e.getMessage());
                System.exit(1);
            }
        }

        final String uuid = arguments.get(0);

        String newStatusName = (arguments.size() == 2) ? arguments.get(1) : null;

        AdminClient client = ClientFactory.createAdminClient(dirAddrs, RPCAuthentication.userService,
                sslOptions, new Options());
        try {
            client.start();
        } catch (Exception e) {
            System.err.println("unable to start client, because:");
        }

        ServiceStatus oldStatus = null;
        try {
            oldStatus = client.getOSDServiceStatus(uuid);
        } catch (Exception e) {
            System.err.println(e.getMessage());
            System.exit(1);
        }

        System.out.print("current status: ");
        switch (oldStatus) {
        case SERVICE_STATUS_AVAIL:
            System.out.println("online");
            break;
        case SERVICE_STATUS_TO_BE_REMOVED:
            System.out.println("locked (read-only)");
            break;
        case SERVICE_STATUS_REMOVED:
            System.out.println("removed");
            break;
        }

        if (newStatusName != null) {
            // change status
            ServiceStatus newStatus = null;
            if (newStatusName.equalsIgnoreCase("online")) {
                newStatus = ServiceStatus.SERVICE_STATUS_AVAIL;
            } else if (newStatusName.equalsIgnoreCase("locked")) {
                newStatus = ServiceStatus.SERVICE_STATUS_TO_BE_REMOVED;
                newStatusName = "locked (read-only)";
            } else if (newStatusName.equalsIgnoreCase("removed")) {
                newStatus = ServiceStatus.SERVICE_STATUS_REMOVED;
            } else {
                System.out.println("unknown status name: " + newStatusName
                        + ". Must be 'online', ' locked' or 'removed'");
                System.exit(1);
            }
            try {
                client.setOSDServiceStatus(uuid, newStatus);
            } catch (IOException e) {
                System.err.println("unable to set new status at DIR, because " + e.getMessage());
                System.exit(1);
            }
            System.out.println("status changed to: " + newStatusName);
        }
        client.shutdown();
        System.exit(0);
    }

    private static void error(String message, Map<String, CliOption> options) {
        System.err.println(message);
        System.out.println();
        usage(options);
        System.exit(1);
    }

    private static void usage(Map<String, CliOption> options) {

        System.out.println("usage: xtfs_chstatus [options] <OSD UUID> [online|locked|removed]");
        System.out.println("<OSD UUID> the registered UUID for which the status is supposed to be changed");
        System.out.println(" online - Marks the OSD as online.");
        System.out.println(" locked - Marks the OSD as locked (locked OSDs will not be assigned to new files).");
        System.out.println("removed - Marks the OSD as no longer available. Replicas on this OSD can be replaced by xtfs_scrub.");
        System.out.println();
        System.out.println("Options:");
        utils.printOptions(options);
    }

}
